trio_asyncio/_handles.py

```diff
@@ -81,86 +53,96 @@ def _raise(self, exc):
             context['source_traceback'] = self._source_traceback
         self._loop.call_exception_handler(context)
 
-    def _repr_info(self):
-        info = [self.__class__.__name__]
-        if self._cancelled:
-            info.append('cancelled')
-        if self._callback is not None:
-            info.append(_format_callback_source(self._callback, self._args))
-        if self._source_traceback:
-            frame = self._source_traceback[-1]
-            info.append('created at %s:%s' % (frame[0], frame[1]))
-        if self._scope is not None:
-            info.append('scope=%s' % repr(self._scope))
-        return info
-
-    def _call_sync(self):
-        assert self._is_sync
+
+class AsyncHandle(ScopedHandle):
+    """A ScopedHandle associated with the execution of an async function.
+    If the handle is cancelled, the cancel scope surrounding the async function
+    will be cancelled too. It is also possible to link a future to the result
+    of the async function. If you do that, the future will evaluate to the
+    result of the function, and cancelling the future will cancel the handle too.
+    """
+
+    __slots__ = ("_fut", "_started")
+
+    def __init__(self, *args, result_future=None, **kw):
+        super().__init__(*args, **kw)
+        self._fut = result_future
+        self._started = trio.Event()
+        if self._fut is not None:
+            @self._fut.add_done_callback
+            def propagate_cancel(f):
+                if f.cancelled():
+                    self.cancel()
+
+    async def _run(self):
+        sniffio.current_async_library_cvar.set("trio")
+        self._started.set()
         if self._cancelled:
             return
-        self._run()
 
-    if sys.version_info >= (3, 7):
-
-        async def _call_async(self, task_status=trio.TASK_STATUS_IGNORED):
-            assert not self._is_sync
-            if self._cancelled:
-                return
-            task_status.started()
+        def report_exception(exc):
+            if not isinstance(exc, Exception):
+                # Let BaseExceptions such as Cancelled escape without being noted.
+                return exc
             try:
-                with trio.CancelScope() as scope:
-                    self._scope = scope
-                    if self._is_sync is None:
-                        await self._context.run(_set_sniff, self._callback, self)
```

Comment:
Note that Context.run() is a synchronous function, so this old code ran the "call async function to get coroutine object" bit in self._context and then awaited the coroutine outside self._context. The new version applies the desired context when running start_soon on the _run task (in BaseTrioEventLoop), so that it gets copied for use in the new task.

Reply:
Makes sense to me ;-)